### PR TITLE
make ws admin can be aware of anything happened in its workspace.

### LINKED ISSUE #2648: missing audits in multi-tenant space
**Describe the Bug**
`ray` is the admin of the ws, `ws-self-provisioner1` is the self-provisioner of the ws. Use this user to create `demo-project1`. Now the default platform `admin` can see the audit of the project creation, but `ray` and `ws-self-provisioner1` can't. Please also check logging/events to see if having the same problem.
Screen Shot 2020-07-26 at 9 29 10 PM
Screen Shot 2020-07-26 at 9 29 40 PM
Screen Shot 2020-07-26 at 9 29 59 PM

**Versions Used**
KubeSphere: 3.0.0-dev
Kubernetes: 1.18.6

### PR DESCRIPTION
**Pull request by wanjunlei**
Signed-off-by: wanjunlei <wanjunlei@yunify.com>

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #2648

### CODE REVIEW

File: pkg/models/tenant/tenant.go [hunk 1]
```diff
@@ -607,6 +607,40 @@ func (t *tenantOperator) listIntersectedNamespaces(workspaces, workspaceSubstrs,
 	return iNamespaces, nil
 }
 
+// listIntersectedWorkspaces returns a list of workspaces that MUST meet ALL the following filters:
+// 1. If `workspaces` is not empty, the workspace SHOULD be one of the specified workpsaces.
+// 2. Else f `workspaceSubstrs` is not empty, the workspace SHOULD be contains one of the specified substrings.
```
Comment (benjaminhuo) [inline pkg/models/tenant/tenant.go:612]: Else f => Else if

File: pkg/models/tenant/tenant.go [hunk 2]
```diff
@@ -828,6 +871,28 @@ func (t *tenantOperator) Auditing(user user.Info, queryParam *auditingv1alpha1.Q
 			namespaceCreateTimeMap[ns.Name] = ns.CreationTimestamp.Time
 		}
 	}
+
```
Comment (benjaminhuo) [inline pkg/models/tenant/tenant.go:874]: Maybe we can add similar comments like the above ns section

File: pkg/models/tenant/tenant.go [hunk 3]
```diff
@@ -607,6 +607,40 @@ func (t *tenantOperator) listIntersectedNamespaces(workspaces, workspaceSubstrs,
 	return iNamespaces, nil
 }
 
+// listIntersectedWorkspaces returns a list of workspaces that MUST meet ALL the following filters:
+// 1. If `workspaces` is not empty, the workspace SHOULD be one of the specified workpsaces.
+// 2. Else f `workspaceSubstrs` is not empty, the workspace SHOULD be contains one of the specified substrings.
+// 3. Else, return all workspace in the cluster.
+func (t *tenantOperator) listIntersectedWorkspaces(workspaces, workspaceSubstrs []string) ([]*tenantv1alpha1.Workspace, error) {
```
Comment (benjaminhuo) [inline pkg/models/tenant/tenant.go:614]: Do we have unit test code for this new function?

File: pkg/simple/client/auditing/elasticsearch/elasticsearch.go [hunk 4]
```diff
@@ -322,6 +322,23 @@ func parseToQueryPart(f *auditing.Filter) interface{} {
...
+						"range": map[string]interface{}{
+							"RequestReceivedTimestamp": map[string]interface{}{
+								"gte": v,
+							},
+						},
+					}},
+				},
+			})
+		}
+
```
Comment (junotx) [inline pkg/simple/client/auditing/elasticsearch/elasticsearch.go:341]: @wanjunlei the relationship between ws  and upper ns condition is or, which may cause problems. For example, an user is a ws admin who queries on an ns and may still see the entire ws.

File: pkg/models/tenant/tenant.go (hunk 3 above)
Comment (wanjunlei) [reply pkg/models/tenant/tenant.go:614]: The logic of this function is very simple, list all workspace, compare with the elements in workspaces and workspaceSubstrs in turn, return results that meet the requirements

File: pkg/simple/client/auditing/elasticsearch/elasticsearch.go [hunk 5]
```diff
@@ -360,6 +360,36 @@ func parseToQueryPart(f *auditing.Filter) interface{} {
 		return &bi
 	}
 
+	if len(f.ObjectRefNamespaces) > 0 {
+		if bi := shouldBoolbody("match_phrase_prefix", "ObjectRef.Namespace.keyword",
```
Comment (benjaminhuo) [inline pkg/simple/client/auditing/elasticsearch/elasticsearch.go:364]: match_phrase_prefix should be change to match_phrase to be more accurate for both namespace and workspace query

File: pkg/apiserver/auditing/types.go [hunk 6]
```diff
@@ -37,16 +41,18 @@ type Auditing interface {
...
 }
 
-func NewAuditing(lister v1alpha1.WebhookLister, url string, stopCh <-chan struct{}) Auditing {
+func NewAuditing(informers informers.InformerFactory, url string, stopCh <-chan struct{}) Auditing {
 
 	a := &auditing{
-		lister: lister,
-		cache:  make(chan *auditv1alpha1.EventList, DefaultCacheCapacity),
+		lister:         informers.KubeSphereSharedInformerFactory().Auditing().V1alpha1().Webhooks().Lister(),
+		resourceGetter: resourcesv1alpha3.NewResourceGetter(informers),
```
Comment (zryfish) [inline pkg/apiserver/auditing/types.go:54]: It's not necessary to create a resourceGetter since just devops needed, we'd better create devopsGetter only

**Pull request merged**